pkg/channel/consolidated/status/status.go

```diff
@@ -0,0 +1,514 @@
+	PodIPs  sets.String
+	PodPort string
+	Port    string
+	URL     *url.URL
+}
+
+// ProbeTargetLister lists all the targets that requires probing.
+type ProbeTargetLister interface {
+	// ListProbeTargets returns a list of targets to be probed
+	ListProbeTargets(ctx context.Context, ch messagingv1beta1.KafkaChannel) (*ProbeTarget, error)
+}
+
+// Manager provides a way to check if an Ingress is ready
+type Manager interface {
+	IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel, sub eventingduckv1.SubscriberSpec) (bool, error)
+}
+
+// Prober provides a way to check if a VirtualService is ready by probing the Envoy pods
+// handling that VirtualService.
+type Prober struct {
+	logger *zap.SugaredLogger
+
+	// mu guards targetStates and podContexts
+	mu           sync.Mutex
+	targetStates map[types.UID]*targetState
+	podContexts  map[string]cancelContext
+
+	workQueue workqueue.RateLimitingInterface
+
+	targetLister ProbeTargetLister
+
+	readyCallback func(messagingv1beta1.KafkaChannel, eventingduckv1.SubscriberSpec)
+
+	probeConcurrency int
+
+	opts []interface{}
+}
+
+// NewProber creates a new instance of Prober
+func NewProber(
+	logger *zap.SugaredLogger,
+	targetLister ProbeTargetLister,
+	readyCallback func(messagingv1beta1.KafkaChannel, eventingduckv1.SubscriberSpec), opts ...interface{}) *Prober {
+	return &Prober{
+		logger:       logger,
+		targetStates: make(map[types.UID]*targetState),
+		podContexts:  make(map[string]cancelContext),
+		workQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewMaxOfRateLimiter(
+				// Per item exponential backoff
+				workqueue.NewItemExponentialFailureRateLimiter(50*time.Millisecond, 30*time.Second),
+				// Global rate limiter
+				&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(50), 100)},
+			),
+			"ProbingQueue"),
+		targetLister:     targetLister,
+		readyCallback:    readyCallback,
+		probeConcurrency: probeConcurrency,
+		opts:             opts,
+	}
+}
+
+func computeHash(sub eventingduckv1.SubscriberSpec) ([sha256.Size]byte, error) {
+	bytes, err := json.Marshal(sub)
+	if err != nil {
+		return [sha256.Size]byte{}, fmt.Errorf("failed to serialize Subscription: %w", err)
+	}
+	return sha256.Sum256(bytes), nil
+}
+
+func (m *Prober) checkReadiness(state *targetState) bool {
+	consumers := int32(state.initialCount)
+	partitions := state.ch.Spec.NumPartitions
+	m.logger.Debugw("Checking subscription readiness",
+		zap.Any("initial probed consumers", consumers),
+		zap.Any("channel partitions", partitions),
+		zap.Any("ready consumers", state.readyCount.Load()),
+	)
+	if consumers > partitions {
+		return state.readyCount.Load() == partitions
+	} else {
+		return state.readyCount.Load() == consumers
+	}
+}
+
+func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel, sub eventingduckv1.SubscriberSpec) (bool, error) {
+	subscriptionKey := sub.UID
+	logger := logging.FromContext(ctx)
+
+	bytes, err := computeHash(sub)
```

Comment:
Why computing hash? Just using `sub.UID` + `sub.Generation` should be fine and be unique